src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs

```diff
@@ -67,18 +65,27 @@ public BuildEventArgsReader(BinaryReader binaryReader, int fileFormatVersion)
 
         public void Dispose()
         {
-            if (stringStorage != null)
-            {
-                stringStorage.Dispose();
-                stringStorage = null;
-            }
+            stringStorage.Dispose();
         }
 
+        /// <summary>
+        /// An event that allows the subscriber to be notified when a string is read from the binary log.
+        /// Subscriber may adjust the string by setting <see cref="StringReadEventArgs.StringToBeUsed"/> property.
+        /// The passed event arg can be reused and should not be stored.
+        /// </summary>
+        public event Action<StringReadEventArgs>? StringReadDone;
```

Comment:
Are Actions the right events here? It might be for this one, but for StringEncountered definitely, I would think a Func<string, string> is most natural, as that gives anyone providing a value for it the opportunity to inspect the string being read, modify it as they wish, and return the modified result.

One implementation, for instance, might be changing some concerning string to a pre-specified obfuscated value and leaving everything else the same. Another might just be the identity function. A third could take every string and hash it. I'm not entirely clear on how the user could use StringEncountered currently to actually change the log produced is what I'm saying, I guess.

Reply:
This results from the decision to use events. Funcs are not good pattern for events. The recommended way of passing values out is via wrapping that in the settable property in the arg of event - see e.g. https://stackoverflow.com/a/1210059/2308106 (by Dustin :-))

So you'd e.g.:
```C#
reader.StringReadDone += args => args.StringToBeUsed = args.StringToBeUsed.Replace(myPwd, "***");
```